rust/azure_iot_operations_connector/src/destination_endpoint.rs

```diff
@@ -3,41 +3,366 @@
+        dataset_definition: &Dataset,
+        inbound_endpoint_name: &str,
+        default_destination: Option<&Destination>,
+        connector_context: Arc<ConnectorContext>,
+    ) -> Result<Self, azure_device_registry::ConfigError> {
         // Create a new forwarder
-        Self {
-            message_schema_uri: Arc::new(RwLock::new(None)),
-        }
+
+        // If no destination is specified in the dataset definition, use the default dataset destination
+        let destination = if dataset_definition.destinations.is_empty() {
+            default_destination.ok_or(azure_device_registry::ConfigError {
+                code: None,
+                details: None,
+                inner_error: None,
+                message: Some("Asset must have default dataset destination if dataset doesn't have destination".to_string()),
+            })?.clone()
+        } else {
+            // for now, this vec will only ever be length 1
+            let definition_destination = &dataset_definition.destinations;
+            Destination::new_dataset_destination(
+                definition_destination,
+                inbound_endpoint_name,
+                &connector_context,
+            )?
+            .expect("Presence of destination already validated")
+        };
+
+        Ok(Self {
+            message_schema_reference: Arc::new(RwLock::new(None)),
+            destination,
+            connector_context,
+        })
     }
 
+    /// Forwards [`Data`] to the destination
+    /// Returns once the message has been sent successfully
+    ///
     /// # Errors
-    /// TODO
-    #[allow(clippy::unused_async)]
-    pub async fn send_data(&self, _data: Data) -> Result<(), String> {
+    /// [`struct@Error`] of kind [`MissingMessageSchema`](ErrorKind::MissingMessageSchema)
+    /// if the [`MessageSchema`] has not been reported yet. This is required before forwarding any data
+    ///
+    /// [`struct@Error`] of kind [`DataValidationError`](ErrorKind::MqttTelemetryError)
+    /// if the [`Data`] isn't valid.
+    ///
+    /// [`struct@Error`] of kind [`BrokerStateStoreError`](ErrorKind::BrokerStateStoreError)
+    /// if the destination is `BrokerStateStore` and there are any errors setting the data with the service
+    ///
+    /// [`struct@Error`] of kind [`MqttTelemetryError`](ErrorKind::MqttTelemetryError)
+    /// if the destination is `Mqtt` and there are any errors sending the message to the broker
+    ///
+    /// # Panics
+    /// if the message schema reference mutex has been poisoned, which should not be possible
+    pub(crate) async fn send_data(&self, data: Data) -> Result<(), Error> {
         // Forward the data to the destination
-        Ok(())
+        match &self.destination {
+            Destination::BrokerStateStore { key } => {
+                if self
+                    .connector_context
+                    .state_store_client
+                    .set(
+                        key.clone().into(),
```

Comment:
Outside of the scope of this PR, but does the fact that you're having to `clone` here suggest that these arguments should be `&[u8]` and not `Vec<u8>`?

Reply:
Not necessarily - the reason that set() takes a Vec<u8> is because there would have to be a clone internally in that function to convert an &[u8] to be the Vec<u8> down the line, so taking it in as a Vec<u8> can potentially reduce a clone if the input wouldn't be used again (not the case here, but if the key were to be unique per call, then that would be the case).

Here we need a clone because the value for the key doesn't change, we get it from a saved definition for each forward, so this is the most optimal way to manage the data, since there needs to be a discrete copy of the value per set() call

Reply:
If that doesn't clear this up for you, ping me - this is a big discussion we've had in the past because it's not always obvious at the surface level